### DISCUSSION TITLE
Homebrew fails to run on Ubuntu 23.04

### DISCUSSION THREAD

**Discussion by nirvdrum** [Homebrew/discussions#5050] [Linux]
### Additional Info

- [ ] I am using Windows Subsystem for Linux

### Output of `brew config`

```shell
/usr/lib/ruby/vendor_ruby/os.rb:7:in `<top (required)>': OS is not a class (TypeError)
/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/macos_version.rb:168: previous definition of OS was here
	from /home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/global.rb:78:in `require'
	from /home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/global.rb:78:in `<top (required)>'
	from /home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/brew.rb:18:in `require_relative'
	from /home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/brew.rb:18:in `<main>'
```
### Output of `brew doctor`

```shell
/usr/lib/ruby/vendor_ruby/os.rb:7:in `<top (required)>': OS is not a class (TypeError)
/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/macos_version.rb:168: previous definition of OS was here
	from /home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/global.rb:78:in `require'
	from /home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/global.rb:78:in `<top (required)>'
	from /home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/brew.rb:18:in `require_relative'
	from /home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/brew.rb:18:in `<main>'
```

### Description of issue

I'm running on a brand new Ubuntu 23.04 installation. I'll try again with Ubuntu 23.10. The system Ruby on this machine is 3.1.2p20, which is not EOL yet.

1. Create Ubuntu 23.04 machine.
2. Install pre-requisites: `sudo apt-get install build-essential procps curl file git`
3. Attempt to install Homebrew: `/bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"`

It looks to me like the vendor_ruby/os.rb file is conflicting with Homebrew. The former defines a class named `OS`, while the latter defines a module named `OS`. I don't know where that vendor_ruby file came from. It's basically a brand new machine so I'm guessing it got pulled in from a system package.

If I look at the `$LOAD_PATH`, I see:

```
["/usr/local/lib/site_ruby/3.1.0",
 "/usr/local/lib/x86_64-linux-gnu/site_ruby",
 "/usr/local/lib/site_ruby",
 "/usr/lib/ruby/vendor_ruby/3.1.0",
 "/usr/lib/x86_64-linux-gnu/ruby/vendor_ruby/3.1.0",
 "/usr/lib/ruby/vendor_ruby",
 "/usr/lib/ruby/3.1.0",
 "/usr/lib/x86_64-linux-gnu/ruby/3.1.0"]
```

Homebrew uses `require "os"` in a couple of locations and my best guess is it's supposed to be loading Homebrew's own [os.rb](https://github.com/Homebrew/brew/blob/ff5cc646993e8fb86ab753c1cefec641af195d63/Library/Homebrew/os.rb). Replacing the `require` with a `require_relative` appears to fix the problem for me.

I can work on a PR, but is there a reason the project is using `require` rather than `require_relative`? The latter is safer and faster if you're looking to load files within your own project. But, using `require` allows for patching code by manipulating the `$LOAD_PATH`. I don't know if that's a use case we need to preserve.

**Comment by Bo98** [comment]
We configure the LOAD_PATH here: https://github.com/Homebrew/brew/blob/5a09a06512fae4c434503edf2f236d379f5bb70c/Library/Homebrew/standalone/init.rb#L42-L46

It seems the `HOMEBREW_LIBRARY_PATH` line should be `unshift` rather than `push`?

**Reply by nirvdrum** [reply]
That fix makes sense as well. I've applied it locally and as far as I can tell, things are working now. I think there's still a case for using `require_relative` for loading files from within the project, but looking at the file you linked I now have a much better understanding of why `$LOAD_PATH` is directly manipulated.

Your suggestion hints at a way to write a regression test as well: we could put a bad file with the same name as a Homebrew file in a directory at the front of the `$LOAD_PATH`. But, I don't know if that's the sort of thing we'd automate. For the most part, if it breaks, you'll hear about it from people.

**Reply by SMillerDev** [reply]
Anything we can automate is worth automating IYAM.
Since you seem to have a solid grasp of the problem, could you make a pull request for at least the fix?

**Reply by nirvdrum** [reply]
I have a good grasp on the problem, less so on the test suite. Where should I pet a test like this? I don't see any existing tests for the `$LOAD_PATH` modifications. I could throw it any one of the CLI tests, I guess, since it impacts the application start-up. But, I'd appreciate any guidance you can provide.

**Reply by SMillerDev** [reply]
I'm not usually involved in that part of brew, but if you put it in a PR wherever you can think of I'm sure someone with more experience will express their opinion

**Reply by nirvdrum** [reply]
Thanks. I was hoping to avoid lengthy round-trips through the review phase for what amounts to a one-line change. But, I'll just open a PR and see what happens.

**Reply by alichtman** [reply]
I opened https://github.com/Homebrew/brew/pull/16479 after running into this issue today.

**Reply by nirvdrum** [reply]
Thanks. Life came up and I haven't had the time to work out the tests yet. It's probably straightforward, but diving into a new code base involves some ramp up.
